woocommerce/wc-admin
# Fix/7358 stock api performance improvement

Fixes #7358 

This PR attempts to improve `low in stock` products REST API endpoint performance.

It adds a new endpoint: `/wp-json/wc-analytics/products/low-in-stock?page=1&per_page=5`

Original query: [gist](https://gist.github.com/moon0326/683fa74d8c606aa60f2ce8ed3b0d07db)
New query: [gist](https://gist.github.com/moon0326/bb49d48b257a1edaa22172759d604c74)

EXPLAIN of the original query (based on a store with 4k products):
![Screen Shot 2021-07-17 at 10 15 59 PM](https://user-images.githubusercontent.com/4723145/126056605-1b3c525e-f556-4c5c-9e04-c31e7f3de781.jpg)

EXPLAIN of the new query (based on a store with 4k products):

![Screen Shot 2021-07-17 at 10 16 28 PM](https://user-images.githubusercontent.com/4723145/126056612-e0dac81d-431a-4c4b-aea9-a398bb09310f.jpg)

In a store with about 10k products, the original query exceeds 30 seconds and times out. The new query takes around 200ms on average. The total query execution time should be 200ms (avg) + 60~80ms (wc_product_meta_lookup query). The query execution can be less than 100ms if the site is using sitewide low stock threshold only.     

### Changes

**The query selects `wc_product_meta_lookup` first then join `wp_posts`**

- This was suggested by @louwie17. I first thought we would not gain much performance, but realized that a store with a large dataset most likely has a higher # of rows in `wp_posts` than `wc_product_meta_lookup`. In a store with 10k products, this change showed 3x performance gain. This is the primary reason why I'm adding a new endpoint since modifying the existing endpoint to use `wc_product_meta_lookup` is quite tricky or nearly impossible without changing almost all of it.

- MySQL always uses `wp_posts` first with the original query, but MariaDB uses `wp_wc_order_product_lookup` depending on the # of rows in each table. This is optimized by the query optimizer automatically. If the query optimizer uses `wp_wc_order_product_lookup` internally anyway, it's safe to select `wp_wc_order_product_lookup` first.

**does not join wp_postmeta if the store is using sitewide stock threshold only**

If a store does not have any custom stock threshold set on a product then we don't have to join `wp_postmeta` table. You can see the method [here](https://github.com/woocommerce/woocommerce-admin/pull/7377/files#diff-d5ee2ec2a16bb74f2d14349119970c8a1725570fe70d1721523df9d40b0ad109R169).

The count query is an extra query, but the execution time is less than 0.1 ms and it can make the main query 2x+ faster if the store is using sitewide low stock threshold only.

**last_order_date is queried separately**

As you see in the EXPLAIN output of the original query, joining `wp_wc_order_product_lookup` is quite expensive. The estimated # of rows to be examined is 10062. The `last_order_date` is now [queried](https://github.com/woocommerce/woocommerce-admin/pull/7377/files#diff-d5ee2ec2a16bb74f2d14349119970c8a1725570fe70d1721523df9d40b0ad109R111) separately.


The query takes between 60~80ms in a store with the following `wc_order_product_lookup` dataset. This store has 10k products.

![Screen Shot 2021-07-17 at 10 40 22 PMth_](https://user-images.githubusercontent.com/4723145/126057042-700d9ff8-6e33-4b8a-a382-60cd407f1562.jpg)

**order by post_date is removed and we now order by wc_product_meta_lookup only**

Since we now select `wp_wc_order_product_lookup` first, ordering by a field that is not in the first selected table triggers `Using temporary; Using filesort`. These two should be avoided whenever possible for the best performance. 

Since `post_date` never gets updated and product_id is the same as post_id, it's safe to use the id field alone in this case.

### Notes

- I noticed that getting product images is also an expensive process as we query each product in a loop. This should be avoided whenever possible, but I'm leaving it as it is for now. We might want to work on this in a separate PR.

### Detailed test instructions:

[4kproducts.sql.zip](https://github.com/woocommerce/woocommerce-admin/files/6835729/4kproducts.sql.zip)

1. Download the `4kproducts.zip`. This SQL dump includes 4297 products for testing purposes.
2. Start with a fresh installation and DB. Import the SQL dump file. 
3. Navigate to WooCommerce -> Status -> Tools and click [ Regenerate ] button on the `Product lookup tables` row.
4. Confirm you have 9k+ rows in `wp_wc_order_product_lookup` table.
5. Navigate to WooCommerce -> Home and hide the task list then refresh.
6. You should see the activity panel with Orders, Stock, and Reviews.
7. Open inspector -> Network -> choose XHR then search for "low_in_stock". The API endpoint should be `products/low-in-stock`

If you want to compare how the new API performs against the existing API, checkout `b341b62b1d4eb7dcc5f088aa77c530ac893b65ec` commit. This commit still makes low-in-stock APIs to the existing endpoint.
